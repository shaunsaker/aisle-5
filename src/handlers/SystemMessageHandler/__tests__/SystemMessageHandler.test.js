import React from 'react';
import renderer from 'react-test-renderer';
import { View } from 'react-native';

import { SystemMessageHandler } from '..';

// Fixes _bezier is not a function bug
jest.useFakeTimers();

describe('SystemMessageHandler', () => {
  const spies = [];
  const dispatch = jest.fn();
  const children = <View />;
  const systemMessage = 'Test';

  describe('renders', () => {
    it('renders with minimum required props', () => {
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch}>{children}</SystemMessageHandler>,
      );

      expect(component).toMatchSnapshot();
    });

    it('renders the snackbar state', () => {
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch} systemMessage={systemMessage}>
          {children}
        </SystemMessageHandler>,
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    it('should handle setHideSnackbar', () => {
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch}>{children}</SystemMessageHandler>,
      );
      const instance = component.getInstance();

      expect(instance.state.hideSnackbar).toEqual(false);

      instance.setHideSnackbar(true);

      expect(instance.state.hideSnackbar).toEqual(true);
    });

    it('should handle onAnimateOut', () => {
      spies[0] = jest.spyOn(SystemMessageHandler.prototype, 'resetSystemMessage');
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch}>{children}</SystemMessageHandler>,
      );
      const instance = component.getInstance();

      instance.onAnimateOut();

      expect(spies[0]).toHaveBeenCalled();
    });

    it('should handle resetSystemMessage', () => {
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch}>{children}</SystemMessageHandler>,
      );
      const instance = component.getInstance();

      instance.resetSystemMessage();

      expect(dispatch).toHaveBeenCalled();
      expect(dispatch).toMatchSnapshot();
    });

    it('should handle clearTimer', () => {
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch}>{children}</SystemMessageHandler>,
      );
      const instance = component.getInstance();

      instance.clearTimer();
    });

    it('should handle onSnackbarPress', () => {
      spies[0] = jest.spyOn(SystemMessageHandler.prototype, 'clearTimer');
      spies[1] = jest.spyOn(SystemMessageHandler.prototype, 'setHideSnackbar');
      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch}>{children}</SystemMessageHandler>,
      );
      const instance = component.getInstance();

      instance.onSnackbarPress();

      expect(spies[0]).toHaveBeenCalled();
      expect(spies[1]).toHaveBeenCalledWith(true);
    });
  });

  describe('lifecycle methods', () => {
    it('should call setHideSnackbar and resetSystemMessage if systemMessage changed in componentDidUpdate', () => {
      spies[0] = jest.spyOn(SystemMessageHandler.prototype, 'clearTimer');
      spies[1] = jest.spyOn(SystemMessageHandler.prototype, 'setHideSnackbar');

      const component = renderer.create(
        <SystemMessageHandler dispatch={dispatch} systemMessage="Something went wrong">
          <View />
        </SystemMessageHandler>,
      );
      const instance = component.getInstance();

      // Setup
      instance.timer = true;

      component.update(
        <SystemMessageHandler dispatch={dispatch} systemMessage="Same same, but different">
          <View />
        </SystemMessageHandler>,
      );

      expect(spies[0]).toHaveBeenCalled();
      expect(spies[1]).toHaveBeenCalledWith(false);

      // FIXME: How to test this?
      // setTimeout(() => {
      //   expect(spies[1]).toHaveBeenCalledWith(true);

      //   done();
      // }, instance.snackbarDuration);
    });
  });

  afterEach(() => {
    spies.forEach((spy) => {
      if (spy) {
        spy.mockClear();
      }
    });
    dispatch.mockClear();
  });
});
