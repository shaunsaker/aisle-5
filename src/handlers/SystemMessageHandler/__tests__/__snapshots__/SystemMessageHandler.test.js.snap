// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemMessageHandler methods should handle resetSystemMessage 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "type": "RESET_SYSTEM_MESSAGE",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`SystemMessageHandler renders renders 1`] = `<View />`;

exports[`SystemMessageHandler renders renders the snackbar state 1`] = `
Array [
  <View />,
  <View
    style={
      Object {
        "alignItems": "center",
        "left": 0,
        "paddingHorizontal": 15,
        "paddingVertical": 10,
        "position": "absolute",
        "right": 0,
        "top": 82,
        "transform": Array [
          Object {
            "translateY": -200,
          },
        ],
      }
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#FEA116",
            "borderColor": null,
            "borderRadius": 5,
            "borderWidth": 0,
            "elevation": 12,
            "opacity": 1,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
            "shadowColor": "#212121",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.33,
            "shadowRadius": 4,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Muli-SemiBold",
              "fontSize": 15,
              "textAlign": "center",
            }
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>,
]
`;

exports[`SystemMessageHandler renders renders with minimum required props 1`] = `<View />`;
